### PR TITLE
Implementation of Issue #3418 - Auto-wrapping text

### PR DESCRIPTION
**Pull request by sptm**
Implementation of Issue #3418 - Auto-wrapping text.
This is done by adding a new parameter 'wrap' to Text.

When enabled, it incrementally builds a temporary line and re-estimates the dimensions of the line for each word added, inserting new lines when needed.

Currently works for any rotated text, different modes of alignment, and for text that are either labels or titles.

### CODE REVIEW

**Comment by tacaswell** [comment]
Nice.  I don't really understand what is causing the failures though.  I assume it is some issue with freetype versions...

**Comment by OceanWolf** [comment]
@tacaswell Anyway to pull the image out of Travis and compare?

@sptm Do you have any rcParams set different from defaults?  For example ``savefig.bbox``

**Comment by andyniemi** [comment]
@tacaswell Freetype2 is the version I've been using to generate the test images on several computers. Is there a version to conform to?

@OceanWolf  In development, rcParams remained the same. It's been recently updated to the MPL source code though when preparing for PR. Would that make a difference?

@sptm Remove the font test case and check?

**Comment by tacaswell** [comment]
exactly which version of freetype2 matters.  All of the current test images are generated with  freetype 2.5.3-21 on Fedora 21 (per https://github.com/matplotlib/matplotlib/pull/4031#issuecomment-71717495).

The failed images on non-PR builds are uploaded to AWS but there is no way to get the images from failed PR build.s

**Comment by tacaswell** [comment]
Also see https://github.com/matplotlib/matplotlib/pull/4345

**Comment by OceanWolf** [comment]
Oooh, brainwave, as part of the testing infrastructure we could catch the test failure email the graphic to the owner of the PR (or something similar) and then re-raise the error.  See http://stackoverflow.com/questions/3362600/how-to-send-email-attachments-with-python for emailing attachments!

**Comment by mdboom** [comment]
@OceanWolf: That's a good idea, and one I've certainly thought about.  The problem is that to e-mail you need e-mail credentials to some mail transport server.  And to use e-mail credentials in a pull request, you are basically exposing those credentials to anyone who wants to submit a pull request (i.e. anyone with a github account).  It won't be long until someone uses those credentials to just send spam.

What we *do* do is upload the images to an Amazon S3 account, and those images are then publically downloadable.  But we can only do it on the main fork (because the only people who can put code in master are "trusted" people), not in any PR.  As a workaround, of course, an author of a PR can ask someone with developer priviledges to push the source branch on the PR to a temporary branch on the main fork.  That causes Travis-CI to test the branch and upload the failed images to Amazon S3.

**Comment by OceanWolf** [comment]
Hmm, re the security issues of email, I can think of two methods:

1. Have github do the email (I don't know if any sort of api exists between travis and github to do that).

2. A tweak to what happens at the moment, we could have travis send the failed images to a/the server we have access to (e.g. the s3 server) which then sends the image as an attachment to the PR submitter or something.  As the server will process the incoming response from Travis you keep the credentials safely locked away on the server.

**Comment by mdboom** [comment]
For (1) -- you'd still need Github credentials for that, even if that were possible.

For (2) -- where do you store the credentials to the S3 server?  That server would cost real money if someone took it over to share files on or something.

See the problem?  Any build of a PR by necessity has exposes the keys to the castle to anyone who can make a PR.  Doing any sort of network operation (other than read-only downloading) requires those keys.  The workaround involves a trusted person (i.e. a matplotlib developer) to take an affirmative action after reading the PR for anything nefarious.  As long as PRs are open to anyone on the internet, I don't see a way around that...

**Comment by OceanWolf** [comment]
@mdboom Actually no I don't see the problem.

Method 1
-----------
While probably not possible at the moment, I wondered how Travis communicates back to github to tell it that the build has succeeded/failed.  For instance how it seems to work at the moment.
1. When you push to github, github creates a ``webhook`` to Travis telling it about the new files.
2. Travis then somehow replies to github telling it that it has started and providing a link ``details`` to the current build.
3. When Travis has finished the build, it then sends another message back to github about whether the build succeeded or failed and github updates it's merge status appropriately.

I suggest here that we tap into the messaging stream which occurs with step 2. and/or 3. operates to have Travis on failure ask github to message the owner of the PR in the same way that github messages the owner of the PR whenever someone posts a message on the PR.  The secure communication aspect for this seems to exist, I just don't know about the API.  If not, then it seems logical that it should and thus could get implemented.  Of course I may completely misunderstand how the communication occurs between Travis and Github, apologies if I have.

Method 2
------------
You don't need to store the credentials for the S3 server, you send a normal request, http 
or something, I think debian does it the same way with reportbug. the server just behaves as a relay.

1. The S3 server receives for example an http POST request with the image and details of the PR.
2. The S3 server validates that Travis sent the request, either from simple ip address checking, (as a more complicated addition it could talk to Travis using the Travis-API to confirm the request came from Travis).
3. The S3 server sends the email (perhaps with a check to github to get the email address).
4. Profit.

Because we keep all the credentials and perform all of the validation on the S3 server, we never expose any secure information, we only show publicly the information in the HTTP POST which contains only data that exists in the public domain anyway (and we can even encrypt using HTTPS, but I don't think we need to).

**Comment by sptm** [comment]
Sorry for the late reply, the freetype version of the most recent one was 2.4.0.
The next one is generated from 2.4.8, which should work in Travis.

**Comment by tacaswell** [comment]
The problem (I think) is that you are using fonts that are not on travis.  Running this on my system (which has _other_ issues with text-related tests (10 of them fail :disappointed: ) ) I get the following images

fonttext_wrap
fonttext_wrap-expected
fonttext_wrap-failed-diff

**Comment by tacaswell** [comment]
Can you also re-base so that the commits with the bad images are not in the history?  We are trying to keep the size of the repo down as much as possible.

**Comment by sptm** [comment]
So it looks like they don't have Times New Roman... thanks for the info.
I'm going to re-work this test and will do a re-base.

**Comment by tacaswell** [comment]
Iirc, there are licensing issues with it. We could probably install it, but
relying only on the fonts we ship with mpl is a better option.

**Comment by jenshnielsen** [comment]
Looking into the Travis logs there are currently there are 3 missing fonts
```
/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/matplotlib-1.5.dev1-py2.6-linux-x86_64.egg/matplotlib/font_manager.py:1285: UserWarning: findfont: Font family [u'Umpush'] not found. Falling back to Bitstream Vera Sans
  (prop.get_family(), self.defaultFamily[fontext]))
/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/matplotlib-1.5.dev1-py2.6-linux-x86_64.egg/matplotlib/font_manager.py:1285: UserWarning: findfont: Font family [u'Times New Roman'] not found. Falling back to Bitstream Vera Sans
  (prop.get_family(), self.defaultFamily[fontext]))
/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/matplotlib-1.5.dev1-py2.6-linux-x86_64.egg/matplotlib/font_manager.py:1285: UserWarning: findfont: Font family [u'Waree'] not found. Falling back to Bitstream Vera Sans
  (prop.get_family(), self.defaultFamily[fontext]))
```

**Comment by andyniemi** [comment]
It looks like it builds now. Any updates or suggestions regarding the PR?

**Comment by tacaswell** [comment]
@ joferkington Any comments as this is based on your SO answer?

File: lib/matplotlib/text.py [hunk 1]
```diff
@@ -315,18 +318,20 @@ def _get_layout(self, renderer):
 
         # Find full vertical extent of font,
         # including ascenders and descenders:
-        tmp, lp_h, lp_bl = renderer.get_text_width_height_descent('lp',
-                                                         self._fontproperties,
-                                                         ismath=False)
+        tmp, lp_h, lp_bl = renderer.get_text_width_height_descent(
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: can you discard this change?

File: lib/matplotlib/text.py [hunk 2]
```diff
@@ -315,18 +318,20 @@ def _get_layout(self, renderer):
...
+            'lp',
+            self._fontproperties,
+            ismath=False)
         offsety = (lp_h - lp_bl) * self._linespacing
 
         baseline = 0
         for i, line in enumerate(lines):
             clean_line, ismath = self.is_math_text(line)
             if clean_line:
-                w, h, d = renderer.get_text_width_height_descent(clean_line,
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: same with this one?

File: lib/matplotlib/text.py [hunk 3]
```diff
@@ -468,12 +473,12 @@ def set_bbox(self, rectprops):
             bbox_transmuter = props.pop("bbox_transmuter", None)
 
             self._bbox_patch = FancyBboxPatch(
-                                    (0., 0.),
-                                    1., 1.,
-                                    boxstyle=boxstyle,
-                                    bbox_transmuter=bbox_transmuter,
-                                    transform=mtransforms.IdentityTransform(),
-                                    **props)
+                (0., 0.),
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: also this one.

File: lib/matplotlib/text.py [hunk 4]
```diff
@@ -601,6 +715,10 @@ def draw(self, renderer):
 
         renderer.open_group('text', self.get_gid())
 
+        if self.get_wrap():
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: This maybe should be done with a context manager to make sure it is always restored correctly?

**Comment by sptm** [comment]
Newest commits uses a context manager to save the old text. Not sure if this is the correct way, as it's my first time using it.

File: lib/matplotlib/text.py [hunk 5]
```diff
@@ -601,56 +731,58 @@ def draw(self, renderer):
...
-        # draw the FancyBboxPatch
-        if self._bbox_patch:
-            self._draw_bbox(renderer, posx, posy)
-
-        gc = renderer.new_gc()
-        gc.set_foreground(self.get_color())
-        gc.set_alpha(self.get_alpha())
-        gc.set_url(self._url)
-        self._set_gc_clip(gc)
+        with _wrap_text(self) as self:
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: I would not yield this as `self`, that is a bit terrifying.

Comment (sptm) [inline lib/matplotlib/text.py]: I can rename this to `text`? `textobj`?

Comment (tacaswell) [inline lib/matplotlib/text.py]: Yes

**Comment by tacaswell** [comment]
@sptm This definitely needs an entry in `https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new`  This is a feature people have been wanting for a long time and it needs to be advertised!

Putting an example in the gallery would also be a good idea.

**Comment by sptm** [comment]
Documentation added. The example is the same one from the test case.

**Comment by tacaswell** [comment]
Awesome!  Thanks for your work.

**Comment by sptm** [comment]
No problem! Thanks for all the help.

**Comment by joferkington** [comment]
@sptm and @tcaswell -  I just wanted to say thanks for the effort in getting this in!  (And with a much nicer implementation than my initial hack, as well.)

**Comment by mghods** [comment]
Thank you for adding this.

**Comment by OceanWolf** [comment]
Shouldn't the ``needs_review`` label have been removed from this when it got merged?

**Comment by jenshnielsen** [comment]
Well there are 218 closed PRs with needs review so whats special about this one. I don't really think thats important since github always does `is:open label:needs_review`

**Comment by OceanWolf** [comment]
Ahh, didn't realise, nothing special, just the only one I have seen, I feel so used to seeing the ``needs_review`` label disappearing once an issue was closed, so much so, I thought it an automatic standard practice thing...

**Comment by jenshnielsen** [comment]
I think the automatic removal is a relatively new thing so it will be missing from older merges.

**Pull request merged**